### PR TITLE
Differentiate UDP and TCP Protocols in Services

### LINKED ISSUE #9207: Differentiate between UDP and TCP services
Currently, we do not differentiate between UDP and TCP services both in the BPF datapath and cilium-agent which means that services of different L4 protocols with the same port number cannot co-exist.

Solving the issue should be thoroughly thought, as we don't want to break the ongoing connections.

### PR DESCRIPTION
**Pull request by aojea**
This is bringing back https://github.com/cilium/cilium/pull/12628 after commented in the community-meeting the concerns about upgrades are well known now and the desire is to have fully conformance Service implementation of Kubernetes.

The code base has changed considerable since this merged and more changes are needed.

> Saving protocol information in the lb(x)_key in bpf, as well as adding protocol information to service maps creation in the lbmap package ensures that translation, receiving, and forwarding will always take the protocol into proper account.
> 
> Refactor cilium to always assume a protocol with a service. Refactor tests to account for service protocols.
> 
> Update Documentation for kubeproxy-free nodeport example Since the output of `cilium service list` has changed the documentation should reflect the change.

Fixes: #9207

```release-note
Cilium differentiate between Protocols on Services for LoadBalancers
```

### CODE REVIEW

**Review by nebril — APPROVED** [review]
Test files LGTM

**Comment by aojea** [comment]
There are a lot of places in the code that assume no protocol , the failures are legit, so far:
- bpf datapath seems ok and most of the code assume protocol
- service caches not ok, lot of places assume no protocol

hold on reviews until I go down this rabbit hole :)

**Comment by aojea** [comment]
I figure out the failures, let's see what the CI says 🤞

**Comment by aojea** [comment]
control plane tests should be fixed now, figuring out the dataplane ones now 🧑‍🏭

**Comment by aojea** [comment]
https://github.com/cilium/cilium/pull/26399/commits/d2d145a08d298c50ac2760192c23d84dcb219754 confirms the datapath changes are the ones causing the problems, is like the reverse nat entries are not found or something

**Comment by aojea** [comment]
the protocol information is not present in the maps
```
root@kind-worker:/home/cilium# bpftool map list | grep serv
37858: hash  name cilium_lb4_serv  flags 0x1
37880: hash  name cilium_lb4_serv  flags 0x1
```

```
bpftool map dump id 37858
key: 0a 60 00 0a 23 c1 00 00  00 00 00 00  value: 00 00 00 00 02 00 00 1c  00 00 00 00
key: 0a 60 00 0a 23 c1 02 00  00 00 00 00  value: 20 00 00 00 00 00 00 1c  00 00 00 00
key: 0a 60 00 0a 00 35 02 00  00 00 00 00  value: 1e 00 00 00 00 00 00 1b  00 00 00 00
key: 0a 60 00 0a 00 35 00 00  00 00 00 00  value: 00 00 00 00 02 00 00 1b  00 00 00 00
key: 0a 60 19 c1 01 bb 00 00  00 00 00 00  value: 00 00 00 00 01 00 00 17  00 10 00 00
key: 0a 60 00 0a 00 35 01 00  00 00 00 00  value: 1d 00 00 00 00 00 00 1b  00 00 00 00
key: 0a 60 b5 bd 00 50 01 00  00 00 00 00  value: 19 00 00 00 00 00 00 18  00 00 00 00
key: 0a 60 00 0a 23 c1 01 00  00 00 00 00  value: 1f 00 00 00 00 00 00 1c  00 00 00 00
key: 0a 60 00 01 01 bb 01 00  00 00 00 00  value: 1a 00 00 00 00 00 00 19  00 00 00 00
key: 0a 60 00 01 01 bb 00 00  00 00 00 00  value: 00 00 00 00 01 00 00 19  00 00 00 00
key: 0a 60 b5 bd 00 50 00 00  00 00 00 00  value: 00 00 00 00 01 00 00 18  00 00 00 00
key: 0a 60 19 c1 01 bb 01 00  00 00 00 00  value: 21 00 00 00 00 00 00 17  00 00 00 00
```

**Comment by aojea** [comment]
found the problem
```
diff --git a/pkg/maps/lbmap/lbmap.go b/pkg/maps/lbmap/lbmap.go
index b5709b1d82..1edc8096f8 100644
--- a/pkg/maps/lbmap/lbmap.go
+++ b/pkg/maps/lbmap/lbmap.go
@@ -74,10 +74,10 @@ func (lbmap *LBBPFMap) upsertServiceProto(p *datapathTypes.UpsertServiceParams,
        backendsOk := ipv6 || !ipv6 && p.NatPolicy != loadbalancer.SVCNatPolicyNat46
 
        if ipv6 {
-               svcKey = NewService6Key(p.IP, p.Port, u8proto.ANY, p.Scope, 0)
+               svcKey = NewService6Key(p.IP, p.Port, u8proto.U8proto(p.Protocol), p.Scope, 0)
                svcVal = &Service6Value{}
        } else {
-               svcKey = NewService4Key(p.IP, p.Port, u8proto.ANY, p.Scope, 0)
+               svcKey = NewService4Key(p.IP, p.Port, u8proto.U8proto(p.Protocol), p.Scope, 0)
                svcVal = &Service4Value{}
        }
```

File: bpf/bpf_sock.c [hunk 1]
```diff
@@ -295,9 +295,11 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 	struct lb4_service *svc;
 	__u16 dst_port = ctx_dst_port(ctx);
 	__u32 dst_ip = ctx->user_ip4;
+	__u32 proto = ctx->protocol;
 	struct lb4_key key = {
 		.address	= dst_ip,
 		.dport		= dst_port,
+		.proto		= (__u8)ctx->protocol,
```
Comment (aojea) [inline bpf/bpf_sock.c:301]: @borkmann is this correct? can I make this assumption to map the protocol from the context to an u8 directly?

**Comment by aojea** [comment]
ok, now it passes all the Kubernetes conformance tests, but there are still failures on

> BPF checks / BPF unit/integration Tests (pull_request)

https://github.com/cilium/cilium/actions/runs/5369880595/jobs/9741636967?pr=26399

**Comment by aojea** [comment]
> ok, now it passes all the Kubernetes conformance tests, but there are still failures on
> 
> >

the problem was the test, fixed, this should be complete green and good to go

**Comment by aojea** [comment]
The failed job is unrelated `IntegrationTests / integration-test (pull_request) Failing after 20m` https://github.com/cilium/cilium/actions/runs/5370117817/jobs/9742035902?pr=26399 , can you please retry it?

> 2023-06-25T14:18:12.5384945Z     testcase.go:302: CiliumNode labels mismatch, expected: map[cilium.io/ci-node:k8s1 kubernetes.io/arch:amd64 kubernetes.io/hostname:cilium-nodes-worker kubernetes.io/os:linux], found: map[cilium.io/ci-node:k8s1 kubernetes.io/arch:amd64 kubernetes.io/hostname:cilium-nodes-worker kubernetes.io/os:linux test-label:test-value]

@aanm @borkmann @tgraf @joestringer this PR makes cilium fully conformance with kubernetes, can you PTAL?

**Comment by brb** [comment]
@aojea Thanks for taking this. How do you handle upgrade/downgrade v1.13 -> your PR -> v1.13, so that existing service connections are not interrupted?

**Comment by aojea** [comment]
> @aojea Thanks for taking this. How do you handle upgrade/downgrade v1.13 -> your PR -> v1.13, so that existing service connections are not interrupted?

Last time we talked about this topic we acknowledged  this was a known issue, so I thought that was going to be explained as disruptive in the release notes ... but I'm open to ideas in this area to implement something that solves it

**Comment by brb** [comment]
For stable features Cilium provides a guarantee that during upgrade no existing flows are interrupted.

The previous attempt was very close to reaching that - https://github.com/cilium/cilium/issues/9207#issuecomment-891784872 (and PR - https://github.com/cilium/cilium/pull/15302).

For this particular problem we could probably tolerate interruptions when doing downgrade. But anyway, I'd suggest to discuss this in the community meeting, and to get a consensus.

**Comment by borkmann** [comment]
> > @aojea Thanks for taking this. How do you handle upgrade/downgrade v1.13 -> your PR -> v1.13, so that existing service connections are not interrupted?

Thanks so much for working on this!

> Last time we talked about this topic we acknowledged this was a known issue, so I thought that was going to be explained as disruptive operations in the release notes ... but I'm open to ideas in this area to implement something that solves it

I think we should at least try to avoid breakage upon upgrade. Could we do the following:

- Add another option to have ANY instead of TCP, UDP, SCTP
- Upon agent restart when we first restore the entries from the maps, from Cilium registers them internally as ANY
- When agent then resyncs with kubeapi-server it would have to leave them as ANY if they were already as such in the map
- For any new service, we populate concrete value TCP, UDP, SCTP, so that over time users will move away from ANY
- From datapath side, we do the lookup first with concrete protocol populated (TCP, UDP, SCTP)
- If there is no service, we retry with proto=0 (ANY) in order to support transition period
- For 1.16 we could remove the 'retry with proto=0 (ANY)' case, so that from that point we only support the one with proto

Wdyt?

**Comment by aojea** [comment]
let me explore both options and we can sync on community meeting, 

I agree that upgrade should not break and being proto 0 for any it seems that we can make the logic smart enough to detect old services and new services

**Comment by nathanjsweet** [comment]
> For 1.16 we could remove the 'retry with proto=0 (ANY)' case, so that from that point we only support the one with proto

That means that you would have to upgrade to 1.15 before you can upgrade to 1.16. Also, it sounds like it would take time for a 1.15 cluster to fully transition to not using ANY. What happens if someone ends up doing a fast upgrade from 1.14 to 1.15?

**Comment by borkmann** [comment]
> What happens if someone ends up doing a fast upgrade from 1.14 to 1.15 to 1.16 without waiting for the full transition?

Good point, yeah, we probably would have to keep ANY around to avoid this.

**Comment by aojea** [comment]
folks I'm going to keep the discussion going here to avoid to spread even more the information

@borkmann @joestringer @nathanjsweet @brb please find attached the doc for discussion https://docs.google.com/document/d/1USB6KnQA2mdJZYxrXn4vvr86hO8V9ItJTgWV7IgHL1A/edit?usp=sharing

**Comment by aojea** [comment]
hmm, it seems this breaks some tests

> ConformanceIngress / ingress-conformance-test (With XDP, native, dedicated, false) (pull_request) Failing after 7m

> * connect to 127.0.0.1 port 31143 failed: Connection refused

**Comment by margamanterola** [comment]
@aojea, are you still working on this? Do you need help with some aspect of this?

**Comment by aojea** [comment]
> @aojea, are you still working on this? Do you need help with some aspect of this?

I'm waiting for a review on the design @margamanterola , I'm very interested in making cillum fully kubernetes conformance

This PR is a prototype demonstrating that the code change is relatively simple, the most complex thing is to implement it without breaking upgrades, so I need first to agree on the implementation design https://docs.google.com/document/d/1USB6KnQA2mdJZYxrXn4vvr86hO8V9ItJTgWV7IgHL1A/edit?usp=sharing

**Comment by hswong3i** [comment]
Kubernetes v1.29.0 e2e conformance test is now blocked by this PR.

Fixes https://github.com/cilium/cilium/issues/29913

**Comment by aojea** [comment]
I don't have much time to work on the PR, but my teammate will follow up soon https://github.com/cilium/cilium/pull/29521

**Comment by chaunceyjiang** [comment]
> I don't have much time to work on the PR, but my teammate will follow up soon https://github.com/cilium/cilium/pull/29521


Hi, @aojea are there any updates on this PR? If not, I'm willing to continue working on this task based on your PR.

**Comment by hswong3i** [comment]
> > I don't have much time to work on the PR, but my teammate will follow up soon #29521
> 
> Hi, @aojea are there any updates on this PR? If not, I'm willing to continue working on this task based on your PR.

Seem like #29521 would be more up-to-date.

**Pull request closed without merging**